src/services/gcs/backend.rs

```diff
@@ -0,0 +1,718 @@
+                } else {
+                    m.set_mode(ObjectMode::FILE);
+                };
+
+                debug!("object {} stat finished: {:?}", &p, m);
+                Ok(m)
+            }
+            StatusCode::NOT_FOUND if p.ends_with('/') => {
+                let mut m = ObjectMetadata::default();
+                m.set_mode(ObjectMode::DIR);
+
+                debug!("object {} stat finished", &p);
+                Ok(m)
+            }
+            _ => {
+                let err =
+                    parse_error_response("stat", args.path(), parse_error_status_code, resp).await;
+                warn!("object {} stat: {:?}", args.path(), err);
+                Err(err)
+            }
+        }
+    }
+
+    #[trace("delete")]
+    async fn delete(&self, args: &OpDelete) -> Result<()> {
+        increment_counter!("opendal_gcs_delete_requests");
+
+        let p = self.get_abs_path(args.path());
+        debug!("object {} delete start", &p);
+
+        let resp = self.delete_object(&p).await?;
+
+        match resp.status() {
+            StatusCode(code) if 200 <= code && code < 300 => {
+                debug!("object {} delete finished", &p);
+                Ok(())
+            }
+            _ => {
+                let err =
+                    parse_error_response("delete", args.path(), parse_error_status_code, resp)
+                        .await;
+                warn!("object {} delete: {:?}", args.path(), err);
+                Err(err)
+            }
+        }
+    }
+
+    #[trace("list")]
+    async fn list(&self, args: &OpList) -> Result<DirStreamer> {
+        increment_counter!("opendal_gcs_list_requests");
+
+        let mut path = self.get_abs_path(args.path());
+        // Make sure list path is endswith '/'
+        if !path.ends_with('/') && !path.is_empty() {
+            path.push('/')
+        }
+        debug!("object {} list start", &path);
+
+        Ok(Box::new(DirStream::new(Arc::new(self.clone()), &path)))
+    }
+
+    fn presign(&self, args: &OpPresign) -> Result<PresignedRequest> {
+        increment_counter!("opendal_gcs_presign_requests");
+
+        let path = self.get_abs_path(args.path());
+
+        // We will not send this request out, just for signing.
+        let mut req = match args.operation() {
+            Operation::Read => self.get_object_request(&path, None, None)?,
+            Operation::Write => {
+                self.insert_object_request(&path, None, isahc::AsyncBody::empty())?
+            }
+            op => {
+                return Err(Error::new(
+                    ErrorKind::Unsupported,
+                    ObjectError::new(
+                        "presign",
+                        &path,
+                        anyhow!("presign for {op} is not supported"),
+                    ),
+                ))
+            }
+        };
+        let url = req.uri().to_string();
+
+        self.signer
+            .sign_query(&mut req, args.expire())
```

Comment:
I remember that gcs has not implemented `sign_query` https://github.com/Xuanwo/reqsign/blob/main/src/services/google/signer.rs